Efficiency should be 0 when there is no path between nodes

arodriguezca opened the pull request:
Currently, when there is no path between nodes, it throws NetworkXNoPath. However, the [author](http://www.uvm.edu/pdodds/files/papers/others/2001/latora2001a.pdf) said it should be 0. Refer to page 2, the paragraph right above formula 1.

dschult commented:
This looks good. 
Could you adjust the docstring so it is fewer than 80 chars wide? (pep8 style).
Thanks!

dschult commented:
If you are up for it, adding a test to tests/test_efficiency.py would be great!

arodriguezca commented:
Style fixed. @dschult, I can do the test. I'll code it later this week.

jarrodmillman commented:
@arodriguezca It would be great if you could add a test in the next few days.  If you have any questions, please let us know.

jarrodmillman commented:
@arodriguezca I would like to merge this before we make a 2.0beta release later this week.  If you prefer, I can add a test for this.  Thanks!

arodriguezca commented:
@jarrodmillman I was disconnected for some days, sorry for the inconveniences. I'll have the test commited by Wednesday night.

jarrodmillman commented:
Great, thanks!

arodriguezca commented:
@dschult @jarrodmillman besides the test for efficiency, I added the case of the global efficiency being 0 when the graph is disconnected (no edges) as it should be according to the paper.

dschult commented:
These tests look good. But they should go in the ```networkx/algorithms/tests/test_efficiency.py``` file rather than the ```networkx/tests/``` directory.   Don't worry about the Codecov errors... those are fixed in another PR.

jarrodmillman commented on networkx/tests/test_efficiency.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        Returns efficiency of two nodes
+        """
+        return nx.efficiency(self.G1, 1, 10)
```
assert_equal(nx.efficiency(self.G1, 1, 10), 0.5)

jarrodmillman commented on networkx/tests/test_efficiency.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        Returns 0 when nodes are disconnected
+        """
+        return nx.efficiency(self.G2, 1, 2)
```
assert_equal(nx.efficiency(self.G2, 1, 2), 0)

jarrodmillman commented on networkx/tests/test_efficiency.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        Test local efficiency
+        """
+        return nx.local_efficiency(self.G1)
```
assert_almost_equal(nx.local_efficiency(self.G1), 0.971146245)

jarrodmillman commented on networkx/tests/test_efficiency.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        In a disconnected graph the efficiency is 0
+        """
+        return nx.local_efficiency(self.G2)
```
assert_equal(nx.local_efficiency(self.G2), 0)

jarrodmillman commented on networkx/tests/test_efficiency.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        test global efficiency
+        """
+        return nx.global_efficiency(self.G1)
```
assert_almost_equal(nx.global_efficiency(G1), 0.525757575)

jarrodmillman commented on networkx/tests/test_efficiency.py:
```diff
@@ -0,0 +1,42 @@
+
```
from nose.tools import assert_equal, assert_almost_equal

arodriguezca commented:
@dschult I didn't notice that there was already one test file there. I merged them.

jarrodmillman commented on networkx/algorithms/tests/test_efficiency.py:
```diff
@@ -1,54 +1,80 @@
 # NetworkX is distributed under a BSD license; see LICENSE.txt for more
 # information.
-"""Unit tests for the :mod:`networkx.algorithms.efficiency` module."""
+# Unit tests for the :mod:`networkx.algorithms.efficiency` module.
```
This should remain a docstring.

jarrodmillman commented on networkx/algorithms/tests/test_efficiency.py:
```diff
@@ -1,54 +1,80 @@
 # test_efficiency.py - unit tests for the efficiency module
 #
-# Copyright 2015 NetworkX developers.
+# Copyright 2017 NetworkX developers.
```
please use ``2015-2017``

jarrodmillman commented on networkx/algorithms/tests/test_efficiency.py:
```diff
@@ -1,54 +1,80 @@
+        return nx.local_efficiency(self.G1)
 
-    """
-    for n in range(10):
```
It looks like the test got dropped.

jarrodmillman replied:
hmm... It also look it had a bug, which is why you probably didn't keep the for loop.

Why not put it back, but use:

```
for n in range(2, 10):
    G = nx.complete_graph(n)
    assert_equal(nx.global_efficiency(G), 1)
```

jarrodmillman commented on networkx/algorithms/tests/test_efficiency.py:
```diff
@@ -1,54 +1,80 @@
+        assert_equal(nx.efficiency(self.G3, 0, 1), 1)
+        assert_equal(nx.efficiency(self.G3, 0, 2), 1 / 2)
 
-    def test_complete_graph(self):
```
You could also do the same thing as I suggested for ``global_efficiency``:
```
for n in range(2, 10):
     G = nx.complete_graph(n)
     assert_equal(nx.local_efficiency(G), 1)
```

jarrodmillman commented on networkx/algorithms/tests/test_efficiency.py:
```diff
@@ -1,54 +1,80 @@
-        # This is the triangle graph with one additional edge.
-        G = nx.lollipop_graph(3, 1)
-        assert_equal(nx.local_efficiency(G), 23 / 24)
+        assert_equal(nx.local_efficiency(self.G5), 23 / 24)
```
This should probably be:
```
assert_equal(nx.local_efficiency(self.G4), 23 / 24)
```

arodriguezca commented:
@jarrodmillman Thanks for all your helpful comments! I fixed the things you pointed out and deleted some functions that I added because they were redundant.
